common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala

```diff
@@ -64,43 +64,34 @@ case class ActivationMessage(override val transid: TransactionId,
   def causedBySequence: Boolean = cause.isDefined
 }
 
-object ActivationMessage extends DefaultJsonProtocol {
-
-  def parse(msg: String) = Try(serdes.read(msg.parseJson))
-
-  private implicit val fqnSerdes = FullyQualifiedEntityName.serdes
-  implicit val serdes = jsonFormat11(ActivationMessage.apply)
-}
-
 /**
  * Message that is sent from the invoker to the controller after action is completed or after slot is free again for
  * new actions.
  */
 abstract class AcknowledegmentMessage(private val tid: TransactionId) extends Message {
   override val transid: TransactionId = tid
-  override def serialize: String = {
-    AcknowledegmentMessage.serdes.write(this).compactPrint
-  }
+  override def serialize: String = AcknowledegmentMessage.serdes.write(this).compactPrint
+  def toJson: JsValue
 }
 
 /**
- * This message is sent from the invoker to the controller, after the slot of an invoker that has been used by the
- * current action, is free again (after log collection)
+ * This message is sent from an invoker to the controller, after the resource slot in the invoke, used by the
+ * corresponding activation, is free again (i.e., after log collection). In some cases, the activation result is
+ * ready and the slot is freed at the same time. In such cases, the completion message carries the result as well.
+ * This is reflected by the of a Right() `response` and the param `result` is set to true.
+ * In some cases the `result` is true but the response is Left() if the message was too large for the message bus.
  */
 case class CompletionMessage(override val transid: TransactionId,
-                             activationId: ActivationId,
-                             isSystemError: Boolean,
+                             response: Either[ActivationId, WhiskActivation],
+                             result: Boolean, // true iff the message is a combined active ack and slot released
```

Comment:
Is this almost the same as:
``` response: Option[Either[ActivationId,WhiskActivation]] ```
? I'm not sure that is more or less clear, but want to make sure I'm not missing something.
In other words - is there any case where result==false, and response==Right[WhiskActivation]?

Reply:
When `result` is false:
1. `Left`: this is the current situation and can occur during split phase notification
1. `Right`: cannot happen, would be a bug (this should be an active ack instead)

When `result` is true:
1. `Right`: this combines the active ack and slot release and occurs during system generate activations for failure scenarios
2. `Left`: like the previous case but occurs when the message is too large to cross the event bus and the sender converts the Right to a Left.

Since there are 3 possible scenarios, `Either` by itself is not enough, and so I chose to encode the extra bit with `result` as a boolean.

Reply:
I don't think `response` as option works - there is always `Some` value (the current case is the `ActivationId`). We need a tri-valued typed. 

An alternative would be `Either[ActivationId, Either[ActivationId, WhiskActivation]` where `Left` represents the split phase and `Right(ActivationId)` and `Right(WhiskActivation)` represents the other two cases.

We can add a `require` in the constructor to also prevent the impossible case.

Reply:
@tysonnorris I thought about this further - another option is to introduce a third type. 

1. Active Ack. (as today)
2. Split phase completion. (what Completion message is today, "result is false" above case 1)
3. Combined Completion. (combines active ack and split phase into one message as this PR does, "result is true" above cases 3 and 4).

There would be 1 active ack + 1 split phase (from the container proxy) and 3 combined completions otherwise (for the error cases).

Reply:
Would this change cause some compatibility issue when there would be a gradual roll out of new invokers in an existing cluster.

So for we do not have any comparability contract for messages exchange via Kafka

Reply:
@rabbah I think it's fine. I don't have a real opinion on any of these options being more clear than the others.  

@chetanmeh I think that is true, but I'm not sure it has been considered for any Message impl changes in the past either? I'm not sure this is explicitly stated anywhere, but I guess it should be, as part of a "system upgrade" documentation?

Reply:
@chetanmeh yes it would since I change one of the property names and the value of the property in the cases now covered by the change may be different (in the corner case, WhiskActivation instead of just ActivationId).

Reply:
> part of a "system upgrade" documentation?

this project doesn't have any yet - are you proposing we start one? ;)

If we introduce a third type, then new invokers could send messages for the bug fix, but an old controller would still ignore them (in fact it would be worse since then the slot release is not recorded)... We could send 2 messages to replicate the split phase notification, but this is wasteful IMO, but then avoid the compatibility issue for a rolling update.

Do you want this PR to address this?

Reply:
@rabbah I do not have the requirement to support rolling updates at the moment. From my side, we can introduce incompatible Kafka message changes requiring that all controllers and invokers need to be updated in parallel.

Reply:
Based on the discussion above, I would say that three different messages sound like the best option. From a data modelling perspective, it's asymmetric that a `Right( WhiskActivation)` "should not happen" in the `CompletionMessage` if `result = false`.

With the split of the active ack into result / completion ack, we somewhat started a renaming. So when introducing a concept with three messages, I suggest to further follow the new naming and use something like `ResultMessage`, `CompletionMessage` and `CombinedResultCompletionMessage`. To preserve consistency, I would update all comments touched by this PR and replace "active ack" with the new terms.

Reply:
the `ack` method will have to change to support this. This is the type of the ack method:
```
  type ActiveAck = (TransactionId, WhiskActivation, Boolean, ControllerInstanceId, UUID, Boolean) => Future[Any]
```

The last parameter in this function indicates if the `slot` is free or not. So it is ambiguous to know inside the method if the result is combined or not from this signature alone since every call already receives a WhiskActivation.

So I'm OK with making the change to the message types (2 to 3) but note that the change will cause further refactoring in the invoker code.

Reply:
@rabbah thanks for your openness to adjust your work to our feedback.

Reply:
@rabbah As you change the current design I would also like to highlight a potential change I need for my activation persister service work

Per my current understand the `sendActiveAck` flow is like

- `isSlotFree` false - active ack/ResultMessage
    - blocking - `WhiskActivation`
    - non-blocking - NONE
- `isSlotFree` true  - result/CompletionMessage
    - blocking - `CompletionMessage`
    - non-blocking - `CompletionMessage`

Later I would like to have a configurable support for sending `WhiskActivation` for non blocking call as well to a custom topic. Just wanted to make you aware on that

Reply:
@chetanmeh I pushed a commit to implement the new type discussed earlier. I did not see your comment earlier. Do the changes help? I think it makes clearer where split-phase acks are being used and where they aren't.